RUFAS/report_generator.py

```diff
@@ -271,3 +296,339 @@ def _prepare_report_data(
+        """
+        Processes and structures a filtered data pool for report generation.
+
+        This method organizes data from a filtered pool based on selected variables and slicing parameters.
+        It caters to different data structures within the pool, ensuring data is formatted appropriately
+        for report inclusion.
+
+        Parameters
+        ----------
+        filtered_pool : Dict[str, pool_element_type]
+            The filtered data pool with each key mapping to its respective data element.
+
+        filter_content : Dict[str, Any]
+            A dictionary containing filter criteria, aggregation instructions, and scalar operation details.
+
+        Returns
+        -------
+        Dict[str, List[Any]]
+            Processed data suitable for report generation, keyed by selected variables.
+
+        Raises
+        ------
+        KeyError
+            If selected_variables is None and the data within the pool requires variable selection.
+        """
+
+        selected_variables = filter_content.get("variables")
+        slice_start = filter_content.get("slice_start", 0)
+        slice_end = filter_content.get("slice_end")
+        report_data: Dict[str, List[Any]] = {}
+
+        for key in filtered_pool.keys():
+            is_data_in_dict = isinstance(filtered_pool[key]["values"][0], dict)
+            if is_data_in_dict and selected_variables is None:
+                raise KeyError(
+                    "Can't generate report, use 'variables' arg to select items from data"
+                )
+            if is_data_in_dict:
+                temp_data = Utility.convert_list_of_dicts_to_dict_of_lists(
+                    filtered_pool[key]["values"][slice_start:slice_end]
+                )
+                for temp_key, temp_values in temp_data.items():
+                    if temp_key not in selected_variables:
+                        continue
+                    if temp_key in report_data:
+                        report_data[temp_key].extend(temp_values)
+                    else:
+                        report_data[temp_key] = temp_values
+            else:
+                report_data[key] = filtered_pool[key]["values"][slice_start:slice_end]
+
+        ReportGenerator._add_constants_data(report_data, filter_content)
+
+        return report_data
+
+    @staticmethod
+    def _add_constants_data(report_data: Dict[str, List[Any]],
+                            filter_content: Dict[str, Any]) -> None:
+        """
+        Add constants to the report data.
+
+        Notes
+        -----
+        An example of a configuration for constants:
+            "constants": [
+                {
+                  "name": "Kilograms to Pounds",
+                  "value": 2.20462
+                },
+                {
+                  "name": "Pounds to Dollars",
+                  "value": 10
+                }
+            ]
```

Comment:
What would you think about changing the way constants are specified to be a dictionary with the key being the name and the value being the value? For example, the above would change to: 
```suggestion
            "constants": {
                "Kilograms to Pounds": 2.20462,
                "Pounds to Dollars": 10
            }
```
This would make the constants section a little more robust because 1) it would not be possible* to pass constants with the same name, and 2) it would not be possible* to forget to add a name or value to any constants (*these would be possible, except the JSON would be invalid - which I think would be a benefit but I can see why this could be considered a drawback).

The drawbacks include the above design being less future-proof (for example, if constants needed to have more elements than just "name" and "value", this design would have to be scrapped), and it also might be a lot of work at this point to change the structure. If either of these are the case, disregard this comment.

Reply:
Right, I do see the merits and drawbacks of either approach. I like the dictionary approach for its succinctness. The array approach allows more info for each constant. Since the code for managing constants is isolated to this part of the code, we can change it later when we know more.

Reply:
I decided to change to your suggestion because with the current use case, the dictionary format is easier for the user and it didn't take too long to make the update.